### ISSUE TITLE
"Bad Message 431. Reason: Request Header Fields Too Large."

### ISSUE THREAD

**Issue by vvaezian** [metabase/metabase#12727]
Some of our company employees who want to try Metabase for the first time by browsing the page `metabase.[OurCompanyName].com`, they get the following error: "Bad Message 431. Reason: Request Header Fields Too Large."

![bad_message](https://user-images.githubusercontent.com/15974735/84835799-3cc4e980-afe9-11ea-93f2-0e685ace7338.png)

This has happened to two out of ~20 users. And for one of them the error didn't happen when they used Internet Explorer but was happening with Chrome and Firefox (I haven't checked the browser of the other person yet).

Since it never happened to me, I cannot reproduce it.
I checked the Metabase logs. This is the related line:  
`WARN http.HttpParser :: Header is too large 8193>8192`

- Metabase version: v0.35.4 (the second incident occurred on this version, but the first one had occurred on a previous version)
- Metabase hosting environment: Elastic Beanstalk
- Metabase internal database: Postgres

It's behind the Firewall in case it matters to the current issue.  
Also we use LDAP for authentication (but this error apparently occurres before loading the login screen. Can confirm that tomorrow.)

**Comment by flamber** [comment]
Hi @vvaezian 
Likely similar to #9386, which is "fixed" by #12683, that allows you to adjust the header size in 0.36.0.
Not sure why it works in IE, but likely because it sends less headers than modern browsers.
You can probably drop some headers with a reverse-proxy as a temporary solution.
Closing for now, but please report back if you're still having problems in 0.36.0 with larger size.

**Comment by vvaezian** [comment]
Thanks. Do you have an estimate on when 0.36.0 will be released?

**Comment by flamber** [comment]
@vvaezian Hopefully in a few weeks - but don't want to give a firm date if something needs more time

**Comment by vvaezian** [comment]
@flamber Some people still have this issue in 38.1. If they clear the cache or use incognito mode the error doesn't happen. So the issue is the cache.
May want to re-open this issue.

**Comment by flamber** [comment]
@vvaezian Have you tried looking at the request to see what is being sent along, before they clear their cache and compare it to when they have cleared it?
Have have you defined https://www.metabase.com/docs/latest/operations-guide/environment-variables.html#mb_jetty_request_header_size to?

**Comment by vvaezian** [comment]
@flamber Below is the error message from MB console when the user got the 431 error message:
```
WARN org.eclipse.jetty.http.HttpParser Header is too large 8193>8192
```
I'll try the `MB_JETTY_REQUEST_HEADER_SIZE` option.

**Comment by thl3538** [comment]
Hello, if the parameters of the report query Get request exceed the maximum length limit of the browser, how to solve this problem, can the Get request be converted into a Post request?

**Comment by colatkinson** [comment]
For anyone else who stumbles across this in the future, there's another possible culprit for bloating headers: `Referer`.

As of this writing (v0.49.3), when Metabase loads an embedded dashboard with token e.g. `/embed/dashboard/a.b.c`, it makes a subsequent API request to `/api/embed/dashboard/a.b.c`.

Browsers will attach a `Referer` header to this API request -- though in most cases, [this will be done only if the header is less than 4096 bytes](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Referer#browser_compatibility). So if your embedding URL is *slightly* less than 4KiB, it will effectively be counted twice against the Jetty header quota: once in the path, and once in the `Referer`. Above this 4KiB limit, the browser will truncate the URL and send only the hostname instead, thus avoiding this issue.

As such, even just doubling the limit to 16384 bytes actually has an outsize effect as these "mid-length" URLs no longer overflow the buffer, and longer URLs are truncated automatically by the browser.

Assuming you're serving Metabase behind a reverse proxy, you can also just strip the `Referer` header out altogether to buy yourself a significant amount of buffer space. Alternatively, you can attach `Referrer-Policy: no-referrer` to responses to achieve much the same.

To my knowledge, there isn't a Metabase setting to control these headers directly, though I'm happy to be corrected!

One other side effect of this is that in incognito mode, in which the `Referer` header is truncated to be origin-only more aggressively, the issue will mysteriously disappear. I assumed this was due to old cookies on the domain or something, but no still just `Referer`.

TL;DR Metabase's header buffer size and modern browser behavior around `Referer` headers don't play nice. In addition to/instead of increasing the buffer size, you can also mess with settings around these headers to avoid the dreaded 431.

**Comment by amr-swalha-wakecap** [comment]
Hello,
In case anyone is still stuck with this issue, we updated the environment for Jetty `MB_JETTY_REQUEST_HEADER_SIZE` inside the docker compose file and we increased the size [[Source](https://www.metabase.com/docs/latest/configuring-metabase/environment-variables#mb_jetty_request_header_size)].

In case you need it, looks like the following:

```
services:
  metabase:
    image: metabase/metabase:latest
    container_name: metabase
    hostname: metabase
    volumes:
      - /dev/urandom:/dev/random:ro
    ports:
      - 3000:3000
    environment:
      MB_JETTY_REQUEST_HEADER_SIZE: 32000
```
